### PR TITLE
Very basic execution of notebooks

### PR DESCRIPTION
**Pull request by jasongrout**
_Very_ basic execution of notebooks and widgets.

### CODE REVIEW

File: src/notebook/plugin.ts [hunk 1]
```diff
@@ -40,20 +51,66 @@ export
...
+  if (isCodeCell(cell)) {
+    let exRequest = {
+      code: cell.input.textEditor.text
+    };
+    let output = cell.output;
+    console.log(`executing`, exRequest)
+    let ex = session.kernel.execute(exRequest);
+    output.clear(false);
+    ex.onIOPub = (msg => {
+      // TODO: not getting an execute_result message
```
Comment (jasongrout) [inline src/notebook/plugin.ts]: @blink1073 - if you have some spare time - it seems that I'm not getting an execute_result message back on the iopub channel, so when I execute the last cell of the example notebook, I'm not seeing the image.  I'm not sure what the issue is and why I'm not getting the execute_result message.

Comment (blink1073) [inline src/notebook/plugin.ts]: I'll look at this first thing tomorrow.

Comment (blink1073) [inline src/notebook/plugin.ts]: The message isn't getting sent by the kernel at all, I set a debug point prior to any validation of the incoming WS message.

Comment (blink1073) [inline src/notebook/plugin.ts]: This works: 

``` python
from IPython.display import Image, display
display(Image('./jupyter.png'))
```

Comment (blink1073) [inline src/notebook/plugin.ts]: Fixed it, the following options must be passed to execute to mimic what is done in the current notebook:

```
{
            silent: false,
            store_history: true,
            stop_on_error: true,
            allow_stdin: true,
        };
```

Comment (blink1073) [inline src/notebook/plugin.ts]: PR Sent.

Comment (jasongrout) [inline src/notebook/plugin.ts]: I'll check tomorrow if this is: (a) not specified in the spec, so it's the server's fault, (b) specified in the spec, but should be handled at the services plugin layer (at least with defaults, or kicking back an error saying your message isn't valid, or (c) my fault for not giving a complete message.

Comment (jasongrout) [inline src/notebook/plugin.ts]: It looks like from the message spec (http://jupyter-client.readthedocs.org/en/latest/messaging.html#execute), the `silent` parameter is what is controlling this, and that should have a default (so should be optional).  I'll look into this more tomorrow.

Comment (jasongrout) [inline src/notebook/plugin.ts]: @blink1073 - it looks like the real problem here is that jupyter-js-services has the wrong defaults: https://github.com/jupyter/jupyter-js-services/blob/master/src/kernel.ts#L636 says the default is `silent=true`, but the docs say the default is `silent=false`.

Comment (jasongrout) [inline src/notebook/plugin.ts]: regardless, I think jupyter-js-services probably shouldn't have to set the defaults. The defaults are encoded in the spec, which I think means that the kernel should implement the defaults, i.e., defaults should be optional members of the messages.

Comment (blink1073) [inline src/notebook/plugin.ts]: I copied the defaults that were actually in the previous source code, happy to update though if we think different ones are better:  https://github.com/jupyter/notebook/blob/master/notebook/static/services/kernels/kernel.js#L736

Comment (jasongrout) [inline src/notebook/plugin.ts]: Okay, this is getting more weird.  I'll continue to dig...

Comment (jasongrout) [inline src/notebook/plugin.ts]: The docs for kernel.js contradict themselves.  https://github.com/jupyter/notebook/blame/master/notebook/static/services/kernels/kernel.js#L700 says silent should default to false, but just below it, https://github.com/jupyter/notebook/blame/master/notebook/static/services/kernels/kernel.js#L710 says silent defaults to true.

It looks like @ellisonbg's commit here changed it: https://github.com/jupyter/notebook/commit/f306423c7ea3bc60735a8bfb4f5f2cc60dc36b97.

Comment (jasongrout) [inline src/notebook/plugin.ts]: I put in a PR to jupyter-js-services here: https://github.com/jupyter/jupyter-js-services/pull/74.  Let's continue the conversation there.

File: src/notebook/plugin.ts [hunk 2]
```diff
@@ -26,6 +32,15 @@ import {
...
 } from 'phosphor-widget';
 
+
+import {
+  CodeCellModel, ICellModel, isCodeCell, BaseCellModel
+} from 'jupyter-js-cells';
+
+import {
+  WidgetManager
+} from './widgetmanager';
```
Comment (blink1073) [inline src/notebook/plugin.ts:42]: Don't forget to check this file in.

Comment (jasongrout) [inline src/notebook/plugin.ts:42]: ah, right, thanks!

**Comment by jasongrout** [comment]
@blink1073 - I think this is probably ready to go in and we can iterate more.

**Comment by jasongrout** [comment]
I thought I could fix the build by adding in the phosphor-widget dependency. However, it's still complaining about what looks like a package version conflict (leading to type errors).  I can't reproduce this locally--everything seems to work fine locally. @blink1073 - does this PR work for you?

**Comment by blink1073** [comment]
It builds fine for me, but I also have node v5.

**Comment by blink1073** [comment]
Ah, try adding phosphor-panel explicitly.

File: src/notebook/plugin.ts (hunk 1 above)
Comment (blink1073) [inline src/notebook/plugin.ts]: This is not still "TODO", right?

Comment (jasongrout) [inline src/notebook/plugin.ts]: right, thanks.

**Comment by jasongrout** [comment]
Thanks - I forgot I used phosphor-panel for my poor-man's gui for notebook execution.

**Comment by jasongrout** [comment]
@blink1073 - I think I rebased this - I just merged in changes from upstream/master.  No conflicts.

**Comment by blink1073** [comment]
I'm still trying on my branch to get the tests to pass.

**Comment by jasongrout** [comment]
Okay.  `npm run test` works great for me, for what it's worth.  There's only one test, right?

**Comment by jasongrout** [comment]
node v4.2.4, npm version 2.14.12

**Comment by blink1073** [comment]
```
$ node -v
v5.0.0
$ npm -v
3.3.9
```

**Comment by blink1073** [comment]
Once @jdfreder updates `jupyter-js-widgets`, the following change will make everything work:

``` diff
diff --git a/example/webpack.conf.js b/example/webpack.conf.js
index c05563c..4a97e5a 100644
--- a/example/webpack.conf.js
+++ b/example/webpack.conf.js
@@ -25,9 +25,14 @@ module.exports = {

     ]
   },
-    externals: {
+  externals: {
       "base/js/namespace": "base/js/namespace",
       "notebook/js/outputarea": "notebook/js/outputarea",
       "services/kernels/comm": "services/kernels/comm"
+  },
+  resolve: {
+    alias: {
+      'requirejs': 'requirejs/require'
+    }
   }
 }
```

**Comment by jasongrout** [comment]
Thanks, I added that resolve key in the latest commit here.

**Comment by jdfreder** [comment]
> Once @jdfreder updates jupyter-js-widgets, the following change will make everything work:

I wouldn't be upset if someone "accidentally" clicked the merge button on that PR :wink:

**Comment by jasongrout** [comment]
The small split PR?

**Comment by jdfreder** [comment]
> The small split PR?

Yeah, it is "small" after all... :smile:

**Comment by blink1073** [comment]
@jasongrout, it looks like we do need `jquery` as a hard dependency.

**Comment by blink1073** [comment]
Can we get this one working before merging https://github.com/ipython/ipywidgets/pull/336?

**Comment by blink1073** [comment]
Ah, nevermind, it is working for me locally once `jquery` is added.

**Comment by blink1073** [comment]
:tada:

**Comment by jdfreder** [comment]
> Can we get this one working before merging ipython/ipywidgets#336?

Yeah, I'm actually waiting on a notebook PR too, so we can't really merge ipython/ipywidgets#336 as quick as I'd like.

**Pull request merged**